File: 11주차/가장먼노드.java
import java.util.*;

public class 가장먼노드 {

    public int solution(int n, int[][] vertex) {
        int answer = 0;

        // 그래프 만들기
        // 인접 리스트로 만듬.
        List<List<Integer>> graph = new ArrayList<>();

        // 0번 사용 안함.
        graph.add(new ArrayList<>());

        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }

        for (int[] node : vertex) {
            int a = node[0];
            int b = node[1];
            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        int[] bfs = BFS(graph);
        int max = Arrays.stream(bfs).max().getAsInt();
        answer = (int) Arrays.stream(bfs).filter(v -> v == max).count();

        return answer;
    }

    public int[] BFS(List<List<Integer>> graph) {
        // 여기다 각 노드의 레벨을 저장한다.
        int[] counts = new int[graph.size()];
        int level = 0;
        // 방문 체크할 배열
        boolean[] isVisit = new boolean[graph.size()];
        Queue<Integer> queue = new LinkedList<>();
        queue.add(1);
        isVisit[1] = true;
        while (!queue.isEmpty()) {
            int qSize = queue.size();
            for (int i = 0; i < qSize; i++) {
                Integer node = queue.poll();
                List<Integer> nodes = graph.get(node);
                for (int n : nodes) {
                    if (!isVisit[n]) {
                        isVisit[n] = true;
                        queue.offer(n);
                        counts[n] = level + 1;
                    }
                }
            }
            level++;
        }

        return counts;
    }

    public static void main(String[] args) {
        int n = 6;
        int[][] vertex = {{3, 6}, {4, 3}, {3, 2}, {1, 3}, {1, 2}, {2, 4}, {5, 2}};

        System.out.println(new 가장먼노드().solution(n, vertex));
    }
}

File: src/11주차/가장먼노드.js
function solution(n, edge) {
  let answer = 0;
  const graph = Array.from(Array(n + 1), () => []);
  const distance = new Array(n + 1).fill(0);
  
  for (const [start, dest] of edge) {
      graph[start].push(dest);
      graph[dest].push(start);
  }
  
  const queue = [1];
  distance[1] = 1;
  
  while (queue.length) {
      const start = queue.shift();
      
      for (const dest of graph[start]) {
          if (!distance[dest]) {
              queue.push(dest);
              distance[dest] = distance[start] + 1;
          }
      }
  }
  
  const max = Math.max(...distance);
  answer = distance.filter((dist) => dist === max).length;

  return answer;
}
